radio/src/targets/common/arm/stm32/trainer_driver.cpp

```diff
@@ -0,0 +1,335 @@
+ *
+ * License GPLv2: http://www.gnu.org/licenses/gpl-2.0.html
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 2 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ */
+
+#include "stm32_pulse_driver.h"
+#include "hal/trainer_driver.h"
+#include "hal.h"
+
+#include "opentx.h"
+
+#if defined(TRAINER_GPIO)
+
+static_assert((TRAINER_OUT_TIMER_Channel == LL_TIM_CHANNEL_CH1 ||
+               TRAINER_OUT_TIMER_Channel == LL_TIM_CHANNEL_CH2 ||
+               TRAINER_OUT_TIMER_Channel == LL_TIM_CHANNEL_CH3 ||
+               TRAINER_OUT_TIMER_Channel == LL_TIM_CHANNEL_CH4) &&
+               __STM32_PULSE_IS_TIMER_CHANNEL_SUPPORTED(TRAINER_OUT_TIMER_Channel),
+              "Unsupported trainer timer output channel");
+
+static_assert(TRAINER_IN_TIMER_Channel == LL_TIM_CHANNEL_CH1 ||
+              TRAINER_IN_TIMER_Channel == LL_TIM_CHANNEL_CH2 ||
+              TRAINER_IN_TIMER_Channel == LL_TIM_CHANNEL_CH3 ||
+              TRAINER_IN_TIMER_Channel == LL_TIM_CHANNEL_CH4,
+              "Unsupported trainer timer input channel");
+#else
+void init_trainer_ppm() {}
+void stop_trainer_ppm() {}
+void init_trainer_capture() {}
+void stop_trainer_capture() {}
+#endif
+
+static void (*_trainer_timer_isr)();
+static const stm32_pulse_timer_t* _trainer_timer;
+
+void init_trainer()
+{
+#if defined(TRAINER_DETECT_GPIO_PIN)
+  LL_GPIO_InitTypeDef pinInit;
+  LL_GPIO_StructInit(&pinInit);
+
+  pinInit.Pin = TRAINER_DETECT_GPIO_PIN;
+  pinInit.Mode = LL_GPIO_MODE_INPUT;
+  pinInit.Pull = LL_GPIO_PULL_UP;
+  LL_GPIO_Init(TRAINER_DETECT_GPIO, &pinInit);
+#endif  
+
+  _trainer_timer = nullptr;
+  _trainer_timer_isr = nullptr;
+}
+
+static inline bool trainer_check_isr_flag(const stm32_pulse_timer_t* tim)
+{
+  switch (tim->TIM_Channel) {
+    case LL_TIM_CHANNEL_CH1:
+      if (LL_TIM_IsEnabledIT_CC1(tim->TIMx) &&
+          LL_TIM_IsActiveFlag_CC1(tim->TIMx)) {
+        LL_TIM_ClearFlag_CC1(tim->TIMx);
+        return true;
+      }
+      break;
+    case LL_TIM_CHANNEL_CH2:
+      if (LL_TIM_IsEnabledIT_CC2(tim->TIMx) &&
+          LL_TIM_IsActiveFlag_CC2(tim->TIMx)) {
+        LL_TIM_ClearFlag_CC2(tim->TIMx);
+        return true;
+      }
+      break;
+    case LL_TIM_CHANNEL_CH3:
+      if (LL_TIM_IsEnabledIT_CC3(tim->TIMx) &&
+          LL_TIM_IsActiveFlag_CC3(tim->TIMx)) {
+        LL_TIM_ClearFlag_CC3(tim->TIMx);
+        return true;
+      }
+      break;
+    case LL_TIM_CHANNEL_CH4:
+      if (LL_TIM_IsEnabledIT_CC4(tim->TIMx) &&
+          LL_TIM_IsActiveFlag_CC4(tim->TIMx)) {
+        LL_TIM_ClearFlag_CC4(tim->TIMx);
+        return true;
+      }
+      break;
+  }
+  return false;
```

Comment:
When there is a readable solution to use the TRAINER_OUT_TIMER_Channel define to select the timer, some CPU cycles could be saved, it may make sense in an interrupt handler
The same applies to the next function.

Reply:
Unfortunately, this is all the HAL LL offers.... each of these functions are inlined and compiled to 1-2 ASM instructions, so that it does not make sense to have function pointers (even slower in that context).

Reply:
Otherwise you end trying to calculate the which register / bit numbers these settings are located, which is hard to do in a portable way, which is what we are trying to achieve when using HAL LL.

Reply:
I could make a macro though to make it more "digestible".

Reply:
As I said, what is there is readable, but the switch could be omitted by using the TRAINER_OUT_TIMER_Channel, though is creates more #if defined foo, of which we have more than enough, that's why this is a maybe

Reply:
When this is your plan the code is good as it is. I fully agree, that this is the way to go.